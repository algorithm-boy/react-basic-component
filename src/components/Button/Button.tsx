import React, { PureComponent, Fragment, MouseEventHandler } from 'react';
import { Style, GeneralProps } from '../global.d';
import { createPropsGetter } from '../../utils';

export interface ButtonProps extends GeneralProps {
  onClick?: (e: React.MouseEvent<HTMLDivElement | HTMLAnchorElement, MouseEvent>) => any;
  disabled?: boolean;
  href?: string;
};

export interface ButtonStates {
  
};

const defaultProps = {
  style: {} as any as Style,
  className: '',
};

type DefaultProps = Readonly<typeof defaultProps>;

const getProps = createPropsGetter(defaultProps);
const opts = {};
class Button extends PureComponent<ButtonProps & DefaultProps, ButtonStates> {
  static defaultProps = defaultProps;
  constructor(props: any) {
    super(props);
  }

  private clickEvent: MouseEventHandler<HTMLDivElement | HTMLAnchorElement> = (event) => {
    const { onClick, disabled } = this.props; 
    if (disabled) {
      return;
    }
    onClick && onClick(event);
  }

  public buttonRender = () => {
    const { style, children, className, disabled } = getProps(this.props);
    
    return <Fragment>
      <div
        className={disabled ? `default disabled ${className}` : `default ${className}`}
        style={style}
        onClick={this.clickEvent}
        {...opts}
      >
        {children}
      </div>

      <style>
        {
          !className && `.default{
          width:100px;
          height:40px;
          border:1px solid #222;
          line-height:40px;
          text-align: center;
          }
          .disabled{
            color:#ccc;
            cursor:not-allowed;
          }
          
          `
        }
      </style>
    </Fragment>;
  }

  public linkRender = () => {
    const { className, style, children, href, disabled } = getProps(this.props);
    return <Fragment>
      <a
        className={disabled ? `default ${className} disabled` : `default ${className}`}
        style={style}
        href={disabled ? 'javascript:void(0)': href}
        onClick={this.clickEvent}
        {...opts}
      >
        {children}
      </a>

      <style>
        {
          !className && `.default{
            display:inline-block;
            width:100px;
            height:40px;
            line-height:40px;
            border:1px solid #222;
            text-align: center;
            text-decoration:none;
            color:#222;
          }
          .disabled{
            color:#ccc;
            cursor:not-allowed;
          }`
          
        }
      </style>
    </Fragment>;
  }

  public render() {
    const { href } = this.props;
    if (href) {
      return this.linkRender();
    }
    return this.buttonRender();

  };
};

export default Button;
