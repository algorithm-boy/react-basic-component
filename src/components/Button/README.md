# Button

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| className(!optional) | Button的类名 (The class property for Button) | string | - |
| style(!optional) | Button的行内样式 (The inline style for Button) | object | {} |
| disable(!optional) | Button的属性 (The inline style for Button) | boolean | false |
| href(!optional) | Button的属性 (The inline style for Button) | string | - |
| onClick(!optional) | Button的属性 (The inline style for Button) | MouseEventHandler | - |

## Example

```
<Button
  className={'hfdc-Button'}
  style={{
    width: '360px'
  }}
/>
```

