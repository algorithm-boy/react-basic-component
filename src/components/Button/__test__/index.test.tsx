import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Button from '../index';

configure({ adapter: new Adapter() });

describe('Button', () => {
  const onButtonClick = jest.fn();
  const setupByMount = () => {
    const wrapper = mount(<Button onClick={onButtonClick} />);
    return {
      wrapper
    };
  };
  it('renders correctly', () => {
    const wrapper = render(
      <Button>Test</Button>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('包含一个a标签', () => {
    const wrapper = shallow(
      <Button href="http://www.baidu.com">Link</Button>
    );
    expect(wrapper.find('a').exists()).toBeTruthy()
  });
  it('点击事件',()=>{
    const { wrapper } = setupByMount();
    wrapper.find('div').at(0).simulate('click');
    expect(onButtonClick).toBeCalled()
  })
 
});
