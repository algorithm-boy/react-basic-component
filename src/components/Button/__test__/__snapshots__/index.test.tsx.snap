// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "default ",
    },
    "children": Array [
      Object {
        "data": "Test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "data": ".default{
          width:100px;
          height:40px;
          border:1px solid #222;
          line-height:40px;
          text-align: center;
          }
          .disabled{
            color:#ccc;
            cursor:not-allowed;
          }
          
          ",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "style",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "style",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {},
          "children": Array [
            Object {
              "data": ".default{
          width:100px;
          height:40px;
          border:1px solid #222;
          line-height:40px;
          text-align: center;
          }
          .disabled{
            color:#ccc;
            cursor:not-allowed;
          }
          
          ",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "style",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "style",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "1": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": ".default{
          width:100px;
          height:40px;
          border:1px solid #222;
          line-height:40px;
          text-align: center;
          }
          .disabled{
            color:#ccc;
            cursor:not-allowed;
          }
          
          ",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "style",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {
        "class": "default ",
      },
      "children": Array [
        Object {
          "data": "Test",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {
            "class": "default ",
          },
          "children": Array [
            Object {
              "data": "Test",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "style",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
